store/mongo/datastore_mongo.go

```diff
@@ -70,77 +78,89 @@ type DataStoreMongoConfig struct {
 }
 
 type DataStoreMongo struct {
-	session     *mgo.Session
+	client      *mongo.Client
 	automigrate bool
 }
 
-func NewDataStoreMongoWithSession(session *mgo.Session) store.DataStore {
-	return &DataStoreMongo{session: session}
+func NewDataStoreMongoWithSession(client *mongo.Client) store.DataStore {
+	return &DataStoreMongo{client: client}
 }
 
-func NewDataStoreMongo(config DataStoreMongoConfig) (store.DataStore, error) {
+//config.ConnectionString must contain a valid
+func NewDataStoreMongo(config DataStoreMongoConfig, l *log.Logger) (store.DataStore, error) {
 	//init master session
 	var err error
 	once.Do(func() {
-
-		var dialInfo *mgo.DialInfo
-		dialInfo, err = mgo.ParseURL(config.ConnectionString)
-		if err != nil {
-			return
+		if !strings.Contains(config.ConnectionString, "://") {
+			config.ConnectionString = "mongodb://" + config.ConnectionString
 		}
-
-		// Set 10s timeout - same as set by Dial
-		dialInfo.Timeout = 10 * time.Second
+		clientOptions := options.Client().ApplyURI(config.ConnectionString)
 
 		if config.Username != "" {
-			dialInfo.Username = config.Username
-		}
-		if config.Password != "" {
-			dialInfo.Password = config.Password
+			clientOptions.SetAuth(options.Credential{
+				Username: config.Username,
+				Password: config.Password,
+			})
 		}
 
 		if config.SSL {
-			dialInfo.DialServer = func(addr *mgo.ServerAddr) (net.Conn, error) {
-
-				// Setup TLS
-				tlsConfig := &tls.Config{}
-				tlsConfig.InsecureSkipVerify = config.SSLSkipVerify
-
-				conn, err := tls.Dial("tcp", addr.String(), tlsConfig)
-				return conn, err
-			}
+			tlsConfig := &tls.Config{}
+			tlsConfig.InsecureSkipVerify = config.SSLSkipVerify
+			clientOptions.SetTLSConfig(tlsConfig)
 		}
 
-		masterSession, err = mgo.DialWithInfo(dialInfo)
+		ctx := context.Background()
+		// ctx, _ := context.WithTimeout(context.Background(), DbConnectionTimeout_s*time.Second)
+		clientGlobal, err = mongo.Connect(ctx, clientOptions)
 		if err != nil {
+			if l != nil {
```

Comment:
IMO it would be better to create logger inside this method instead of taking it as an argument and checking if it's nil or not

Reply:
right!